### PR TITLE
Close underlying HID device when done with it

### PR DESCRIPTION
**Pull request by scgbckbone**
On Ubuntu 20.04:
`Linux workstation 5.13.0-27-generic #29~20.04.1-Ubuntu SMP Fri Jan 14 00:32:30 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`
with `Python 3.8.10` I have encountered `Segmentation fault (core dumped)` after every cli command (check added screenshot). It is caused by device not being closed properly.

This PR resolved the above mentioned issue by converting `get_device` function to context manager which executes `ColdcardDevice.close()` on exit. 

Screenshot from 2022-01-27 14-51-40

### CODE REVIEW

**Comment by scgbckbone** [comment]
yes - `dev = dev or get_device()` this line in `real_file_upload` wasn't playing good with context manager - so I changed it to positional so you need to always provide a device

File: ckcc/cli.py [hunk 1]
```diff
@@ -729,107 +731,104 @@ def bip39_passphrase(passphrase, verbose=False):
 @click.option('--path', '-p', default="m/45'", help="Derivation for key (default: BIP45 = m/45')")
 @click.option('--add', '-a', 'just_add', is_flag=True, help='Just show line required to add this Coldcard')
 def enroll_xpub(name, min_signers, path,  num_signers, output_file=None, verbose=False, just_add=False):
-    '''
-Create a skeleton file which defines a multisig wallet.
-
```
Comment (doc-hex) [inline ckcc/cli.py:734]: I think the blank line you removed affects the `--help`` message

File: ckcc/cli.py [hunk 2]
```diff
@@ -938,76 +937,74 @@ def user_auth(psbt_file, next_code=None):
 @click.option('--debug', '-d', is_flag=True, help='Show values used')
 @click.option('--version3', '-3', is_flag=True, help='Support obsolete 3.x.x firmware')
 def user_auth(username, token=None, password=None, prompt=None, totp=None, psbt_file=None, debug=False, version3=False):
-    '''
-Indicate specific user is present (for HSM).
+    """
```
Comment (doc-hex) [inline ckcc/cli.py:940]: would rather you hadn't changed all the `'''` into `"""` ... everywhere

**Comment by doc-hex** [comment]
Thanks for this fix, altho I can't test it... it's a big diff, so will have to review more, and I raised some concerns.

Wondering if `atexit()` might have been easier approach. Also, I consider a bug in OS/libusb/libhid if a segfault occurs like this.

**Comment by scgbckbone** [comment]
yes that was my thinking that something must be broken/outdated in OS libraries. But I still think that closing resources explicitly after we have no use for them is good practice at least. 

Secondary, testing, my second question - how do you test this? only manually ?

**Comment by doc-hex** [comment]
protocol code is part of coldcard/firmware's test cases
no test code yet for CLI stuff, sorry

**Comment by scgbckbone** [comment]
it is tough to evaluate whether changes are good to go or not without any unit tests - the hard part is that we need to communicate with device which sounds more like integration tests. At least some unit tests with mocked CC may be good idea - will look into it.

**Comment by doc-hex** [comment]
sure would love some test cases! use pytest and the Coldcard Simulator as the target (`-x` flag)

**Comment by scgbckbone** [comment]
I do not think that `atexit` is the right tool as the `.close` is instance method on the object

**Comment by scgbckbone** [comment]
What about this piece?

File: ckcc/cli.py [hunk 3]
```diff
@@ -260,135 +267,129 @@ def real_file_upload(fd, blksize=MAX_BLK_LEN, do_upgrade=False, do_reboot=True,
...
 def file_upload(filename, blksize, multisig=False):
-    "Send file to Coldcard (PSBT transaction or firmware)"
+    """Send file to Coldcard (PSBT transaction or firmware)"""
 
     # NOTE: mostly for debug/dev usage.
-    dev = get_device()
+    with get_device() as dev:
 
-    file_len, sha = real_file_upload(filename, blksize, dev=dev)
+        file_len, sha = real_file_upload(filename, blksize, dev=dev)
```
[the remark this answers is not in the document]
Comment (doc-hex) [reply ckcc/cli.py:283]: this bug needs fix

**Comment by doc-hex** [comment]
https://github.com/Coldcard/ckcc-protocol/pull/14/files#diff-cd22cd9db4b43f5f3c467b4e31bbca7a51e749fb4b0b3d24a5c51733dbfa05edR283 needs a fix, and search for similar occurances

**Comment by scgbckbone** [comment]
thanks for pointing that out! fixed

**Pull request merged**